ocaml/xapi/repository_helpers.ml

```diff
@@ -1143,71 +1151,39 @@ let eval_guidances ~updates_info ~updates ~kind =
   |> GuidanceSet.resort_guidances ~kind
   |> GuidanceSet.elements
 
-let parse_line_of_list_updates (buff, acc) line =
-  let get_pkg name_arch epoch_ver_rel repo =
-    let name, arch = Pkg.parse_name_arch name_arch in
-    let epoch, version, release =
-      Pkg.parse_epoch_version_release epoch_ver_rel
-    in
-    let pkg = Pkg.{name; epoch; version; release; arch} in
-    (pkg, repo)
-  in
-  let empty_buff = (None, None, None) in
-  let sep = Re.Str.regexp " +" in
-  match (buff, Re.Str.split sep line) with
-  | _, ["Updated"; "Packages"] ->
-      (empty_buff, acc)
-  | (None, None, None), [name_arch] ->
-      Pkg.parse_name_arch name_arch |> ignore ;
-      ((Some name_arch, None, None), acc)
-  | (None, None, None), [name_arch; epoch_ver_rel] ->
-      Pkg.parse_name_arch name_arch |> ignore ;
-      ((Some name_arch, Some epoch_ver_rel, None), acc)
-  | (None, None, None), [name_arch; epoch_ver_rel; repo] ->
-      Pkg.parse_name_arch name_arch |> ignore ;
-      let pkg = get_pkg name_arch epoch_ver_rel repo in
-      (empty_buff, pkg :: acc)
-  | (None, None, None), _ ->
-      warn "Can't parse '%s'. Not data in buff." line ;
-      (empty_buff, acc)
-  | (Some name_arch, None, None), [epoch_ver_rel] ->
-      ((Some name_arch, Some epoch_ver_rel, None), acc)
-  | (Some name_arch, None, None), [epoch_ver_rel; repo] ->
-      let pkg = get_pkg name_arch epoch_ver_rel repo in
-      (empty_buff, pkg :: acc)
-  | (Some name_arch, None, None), _ ->
-      warn "Can't parse '%s'. name_arch %s is in buff." line name_arch ;
-      (empty_buff, acc)
-  | (Some name_arch, Some epoch_ver_rel, None), [repo] ->
-      let pkg = get_pkg name_arch epoch_ver_rel repo in
-      (empty_buff, pkg :: acc)
-  | (Some name_arch, Some epoch_ver_rel, None), _ ->
-      warn "Can't parse '%s'. name_arch %s and epoch_ver_rel %s are in buff."
-        line name_arch epoch_ver_rel ;
-      (empty_buff, acc)
+let parse_line_of_repoquery acc line =
+  match Astring.String.cuts ~sep:":" line with
+  | [name; epoch'; version; release; arch; repo] -> (
+    try
+      let epoch = Epoch.of_string epoch' in
+      let pkg = Pkg.{name; epoch; version; release; arch} in
+      (pkg, repo) :: acc
+    with e ->
+      warn "epoch %s from repoquery: %s" epoch' (ExnHelper.string_of_exn e) ;
+      (* The error should not block update. Ingore it. *)
+      acc
+  )
   | _ ->
-      warn "Can't parse '%s'" line ;
-      (empty_buff, acc)
+      warn "Can't parse line of repoquery '%s'" line ;
+      acc
 
-let get_updates_from_list_updates repositories =
-  let params_of_list =
+let get_updates_from_repoquery repositories =
+  let params =
     [
-      "-q"
+      "-a"
     ; "--disablerepo=*"
     ; Printf.sprintf "--enablerepo=%s" (String.concat "," repositories)
-    ; "list"
-    ; "updates"
+    ; "--pkgnarrow=updates"
+    ; "--qf"
+    ; "%{name}:%{epoch}:%{version}:%{release}:%{arch}:%{repoid}"
```

Comment:
The parsing of this line relies on the precondition that `:` cannot appear in any of these fields. Is this actually the case? If there is no certainty, characters that I would consider to be less likely to be used in a any of these fields are space, `|`, `<` - or any character that has meaning in a shell because packages would avoid them even when they are not illegal.

Reply:
it would be good to make this and the rpm -qa format consistent: whatever concerns we have for parsing 'repoquery' output the same applies to parsing 'rpm -qa' output, and in fact the query format here and in rpm -qa should use the same field names. So would be good to define this query format as a constant, and then next to it move the function that parses that format: that way it'll be obvious that one needs to be updated when the other one is, and we can share code between rpm -qa and repoquery more easily.

Reply:
The `:` will not appear in any of these fields as it is a special one for epoch. For example: `libpath-utils-2:0.2.1-29.el7.x86_64` means the epoch is `2`. So I think `:` should be safe.
I can't find a document to clarify this definitively :(

Reply:
How about using `:\t` as separator? `/usr/lib/rpm/rpmpopt-4.11.3` has a bunch of queries as examples used by RPM itself internally and separators used there are `:`, `\t`,`|` and ` `.
Even if a package would have a whitespace or : somewhere it is very unlikely it'd be able to have both next to each other, and then you could use `cuts` to split on `":\t"` (of course the danger here is that something would convert tabs to spaces, but it'll be very obvious if that happens)

Reply:
https://github.com/xapi-project/xen-api/blob/a9e9763ad2d2282df5426e0959a14df37ab7f658/ocaml/xapi/repository_helpers.ml#L1049 this is the 2nd occurance of 'rpm -qa' that I'm refering to, and it would be nice to use the same format string there and in repoquery, and same parsing function.

Reply:
Thanks @lindig @edwintorok 
I've appended a commit for this change.